crypto/x509/x509_cmp.c

```diff
@@ -452,7 +452,16 @@ STACK_OF(X509) *X509_chain_up_ref(STACK_OF(X509) *chain)
     ret = sk_X509_dup(chain);
     for (i = 0; i < sk_X509_num(ret); i++) {
         X509 *x = sk_X509_value(ret, i);
-        X509_up_ref(x);
+        if (!X509_up_ref(x)) {
+            int j;
+
+            for (j = 0; j < i; j++) {
+                x = sk_X509_value(ret, j);
+                X509_free(x);
+            }
+            sk_X509_free(ret);
+            return NULL;
+        }
```

Comment:
I guess you can make this much simpler like this:
```
ret = sk_X509_new(X509_cmp);
for (i = 0; i < sk_X509_num(chain); i++) {
    X509 *x = sk_X509_value(chain, i);
    if (!X509_up_ref(x)) {
        sk_X509_pop_free(ret, X509_free);
        return NULL;
    }
    sk_X509_push(ret, x);
}
return ret;
```

Reply:
The sk_X509_new() should probably get the compare function from chain, but we currently don't seem to have a way to get that with public functions.

Reply:
Thanks...that gave me another idea.